src/content/reference/device-os/firmware.md

```diff
@@ -12979,6 +13005,1754 @@ int8_t measurePower() const;
+| Variant::Type::MAP | `VariantMap` |
+
+
+### constructor [Variant class]
+
+{{api name1="Variant::Variant()"}}
+
+You can construct a `Variant` object with a parameter of an explict type to create a variant of that type.
+
+```cpp
+// PROTOTYPES
+Variant();
+Variant(const std::monostate& val);
+Variant(bool val);
+Variant(int val);
+Variant(unsigned val);
+Variant(long val);
+Variant(unsigned long val);
+Variant(long val);
+Variant(unsigned long val);
+Variant(long long val);
+Variant(unsigned long long val);
+Variant(double val);
+Variant(const char* val);
+Variant(String val);
+Variant(VariantArray val);
+Variant(VariantMap val);'
+```
+
+### value() [Variant class]
+
+{{api name1="Variant::value()"}}
+
+The `value` method is a template that returns a reference to the stored value of the variant. The allowable types are listed in [Variant::Type](#variant-type). This can be done to either get or set a value.
+
+```cpp
+// PROTOTYPES
+template<typename T>
+T& value();
+
+template<typename T>
+const T& value() const;
+
+// EXAMPLES
+int value1 = variant1.value<int>();
+bool value2 = variant2.value<bool>();
+String value3 = variant3.value<String>();
+```
+
+
+### type() [Variant class]
+
+{{api name1="Variant::type()"}}
+
+Return the type of the `Variant`. See [Variant::Type](#variant-type) for a list of valid types. There are also [accessors to check for a specific type](#isxxx-variant-class-) such as `isInt()`.
+
+```cpp
+// PROTOTYPE
+Type type() const;
+```
+
+### isXXX() [Variant class]
+
+Returns true if the Variant is of the specified type.
+
+```cpp
+// PROTOTYPES
+bool isNull() const;
+bool isBool() const;
+bool isInt() const;
+bool isUInt() const;
+bool isInt64() const;
+bool isUInt64() const;
+bool isDouble() const;
+bool isNumber() const;
+bool isString() const;
+bool isArray() const;
+bool isMap() const;
+```
+
+
+### toBool() [Variant class]
+
+{{api name1="Variant::toBool()"}}
+
+Returns value of the variant to an `bool` (boolean, `true` or `false`), converting the type if necessary. The original value is left unchanged.
+
+| Source | Result | 
+| :--- | :--- |
+| bool | unchanged |
+| numeric | zero (`0`) &rarr; `false`, non-zero &rarr; `true` |
+| String | `false` &rarr; `false`, 'true' &rarr; "true" |
+| other | conversion fails |
+
+There are two overloads, one that takes an `ok` parameter passed by reference, which is filled in as 
+`true` if the data is already this type or can be converted.
+
+```cpp
+// PROTOTYPES
+bool toBool() const;
+bool toBool(bool& ok) const;
+```
+
+### asBool() [Variant class]
+
+{{api name1="Variant::asBool()"}}
+
+Returns a reference to the value contained in this variant as a `bool`. This can be used to modify the value of the variant.
+
+```cpp
+// PROTOTYPE
+bool& asBool();
+```
+
+This method will convert the type of the variant to bool if necessary, see [toBool](#tobool-variant-class-) for more information. If the conversion is not possible, the method will assert.
```

Comment:
It's not clear what's the difference between `value<bool>()`, `asBool()` and `toBool()`. Perhaps it's worth adding a section about that so that we don't have to document it for each of these methods separately.

- `value<T>()` – returns a reference to the currently held value. Requires it be of the `T` type otherwise would cause undefined behavior.
- `as<T>()` or `asXXX()` – converts the underlying value of the variant to `T` _in place_ and returns a reference to it. Safe to call if the types don't match but requires `T` to be one of the supported types, otherwise would cause a compilation error (not a runtime assert as currently documented).
- `to<T>` or `toXXX()` – converts the underlying value to `T` without modifying the variant and returns the result _by value_. Safe to call if the types don't match, `T` can be arbitrary.

Reply:
Added a new section with this explanation, and linked to it from all of the methods.